sklearn/naive_bayes.py

```diff
@@ -35,6 +35,30 @@
 __all__ = ['BernoulliNB', 'GaussianNB', 'MultinomialNB']
 
 
+def _safe_logprob(p):
+    """
+    Setting log(0) = -inf as -th where `th` is a very larger number.
+
+    Parameters
+    ----------
+    p : array-like or scalar
+        Probabilities values. Must be within [0, 1] range.
+
+    This avoids summation errors, e.g. np.log(0) - np.log(0) = nan, which
+    may happen while calculating dot product of log probability matrices for
+    joint likelihood estimation
+
+    Note that `th` too large will cause overflow during dot product and wrongly
+    estimate the joint likelihood
+    """
+    th = 1e30
+    p = np.asarray(p)
+    if (p > 1).any() or (p < 0).any():
+        raise ValueError('Input `p` must be within [0, 1] range!')
```

Comment:
@agramfort Thanks for your comment. It just happens when some elements contains `inf` (or very close to overflow which leads to `inf`), `np.dot` or `np.array.sum` may result in `nan`. Some simple examples:
```python
In[3]: np.inf - np.inf
Out[3]: nan
In[7]: 1e+308 + 1e+308
Out[7]: inf
In[8]: (1e+308 + 1e+308) - (1e+308 + 1e+308)
Out[8]: nan
```
Since we are calculating log probability, I think it is reasonable to use a large negative number like `-1e30` to replace `log(0) = -inf` when probability is 0. 

> why not clipping systematically?

I'm not sure I understand but do you mean creating wrappers for numpy to clip all values?